### PR TITLE
refactor AI economy conditions

### PR DESCRIPTION
**Pull request by relent0r**
## Description of the changes
This PR replaces some of the default economy condition functions. By default all conditions perform a call to the function aiutilities.lua AIGetEconomyNumbers. This function runs 13 economy functions and does a few other calculations then returns a table. The economy condition then takes the economy property its interested in from this.
For instant build conditions this gets called each time the build condition monitor performs its condition check. This PR removes the excess calls by making the build conditions only call the functions they need for the data required.

## Test setup for the changes
Perform an AI game and record the number of times the economy based functions are called using the built in profiler. Then rerun the same game via replay with the refactored code. Ensure no desync due to errors and compare the number of times the functions are called. The refactored set should be less by a number of thousands over a period of time.

Note : I haven't touched the EconOverTime functions or StorageMax functions as these require extra calculations that are done within the AIGetEconomyNumbers function. There are also a bunch of function calls in the file SorianInstantBuildConditions. Some are custom and others are duplicates, would recommend removing the duplicates and re pointing the builders to the default conditions via a different PR.

This is the number of calls made prior to refactor over a 10 minute game.
FunctionCallsPreRefactor

This is the number of calls made post refactor across the same 10 minute game.
FunctionCallsPostRefactor

### CODE REVIEW

File: lua/editor/EconomyBuildConditions.lua [hunk 1]
```diff
@@ -237,7 +233,7 @@ end
 ---@param EnergyEfficiency integer
 ---@return boolean
 function GreaterThanEconEfficiencyOverTime(aiBrain, MassEfficiency, EnergyEfficiency)
-    if HaveGreaterThanUnitsWithCategory(aiBrain, 0, 'ENERGYPRODUCTION EXPERIMENTAL STRUCTURE') then
+    if HaveGreaterThanUnitsWithCategory(aiBrain, 0, categories.ENERGYPRODUCTION * categories.EXPERIMENTAL * categories.STRUCTURE) then
```
Comment (Garanas) [inline lua/editor/EconomyBuildConditions.lua:236]: You can cache these categories too, as an upvalue. See also: https://github.com/FAForever/fa/blob/deploy/fafdevelop/lua/benchmarking/benchmarks/categories.lua

**Comment by Garanas** [comment]
Good use of the profiler to back up your changes 👍 . The changes appear good to me, @Uveso could you glance over it too?

File: lua/editor/EconomyBuildConditions.lua (hunk 1 above)
Comment (relent0r) [reply lua/editor/EconomyBuildConditions.lua:236]: good point, I'll update to include that.

**Comment by Uveso** [comment]
looks good.

i would use as paragon cat this:
```categories.STRUCTURE * categories.EXPERIMENTAL * categories.ECONOMIC * categories.ENERGYPRODUCTION * categories.MASSPRODUCTION```

only using ENERGYPRODUCTION EXPERIMENTAL STRUCTURE will lead to some buildings that are recognized as paragon but are only modded energy structures like the Black Ops "Philadelphia" (Stellar Generator)
There are also some T4 energy generator mods that will also make problems.

**Comment by relent0r** [comment]
> categories.STRUCTURE * categories.EXPERIMENTAL * categories.ECONOMIC * categories.ENERGYPRODUCTION * categories.MASSPRODUCTION

Ok will update to use this.

**Comment by Garanas** [comment]
Is the check specifically for the paragon? You can also check for the unit id, all unit ids are categories too.

**Comment by Uveso** [comment]
This check should be for paragon like units, not paragon only.

With unit ID this would not recognize units from my "Quantum Resource Generators" mod, "paragon for all" mod etc.

**Pull request merged**